### ISSUE TITLE
Please link executable with '-pie'

### ISSUE THREAD

**Issue by ghost** [cornelisnetworks/opa-fm#11]
When I check opa-fm rpm package with "annocheck". I got these error message.

Hardened: ./usr/lib/opa-fm/bin/config_check: FAIL: Not linked as a position independent executable (ie need to add '-pie' to link command line).
Hardened: ./usr/lib/opa-fm/bin/config_convert: FAIL: Not linked as a position independent executable (ie need to add '-pie' to link command line).
Hardened: ./usr/lib/opa-fm/bin/fm_cmd: FAIL: Not linked as a position independent executable (ie need to add '-pie' to link command line).
Hardened: ./usr/lib/opa-fm/bin/opafmd: FAIL: Not linked as a position independent executable (ie need to add '-pie' to link command line).
Hardened: ./usr/lib/opa-fm/bin/opaxmlextract: FAIL: Not linked as a position independent executable (ie need to add '-pie' to link command line).
Hardened: ./usr/lib/opa-fm/bin/opaxmlfilter: FAIL: Not linked as a position independent executable (ie need to add '-pie' to link command line).
Hardened: ./usr/lib/opa-fm/bin/smpoolsize: FAIL: Not linked as a position independent executable (ie need to add '-pie' to link command line).
Hardened: ./usr/lib/opa-fm/runtime/fe: FAIL: Not linked as a position independent executable (ie need to add '-pie' to link command line).
Hardened: ./usr/lib/opa-fm/runtime/sm: FAIL: Not linked as a position independent executable (ie need to add '-pie' to link command line).
Hardened: ./usr/lib/opa-fm/bin/config_check: FAIL: Not linked as a position independent executable (ie need to add '-pie' to link command line).
Hardened: ./usr/lib/opa-fm/bin/config_convert: FAIL: Not linked as a position independent executable (ie need to add '-pie' to link command line).
Hardened: ./usr/lib/opa-fm/bin/fm_cmd: FAIL: Not linked as a position independent executable (ie need to add '-pie' to link command line).
Hardened: ./usr/lib/opa-fm/bin/opafmd: FAIL: Not linked as a position independent executable (ie need to add '-pie' to link command line).
Hardened: ./usr/lib/opa-fm/bin/opaxmlextract: FAIL: Not linked as a position independent executable (ie need to add '-pie' to link command line).
Hardened: ./usr/lib/opa-fm/bin/opaxmlfilter: FAIL: Not linked as a position independent executable (ie need to add '-pie' to link command line).
Hardened: ./usr/lib/opa-fm/bin/smpoolsize: FAIL: Not linked as a position independent executable (ie need to add '-pie' to link command line).
Hardened: ./usr/lib/opa-fm/runtime/fe: FAIL: Not linked as a position independent executable (ie need to add '-pie' to link command line).
Hardened: ./usr/lib/opa-fm/runtime/sm: FAIL: Not linked as a position independent executable (ie need to add '-pie' to link command line).

This patch fixes this issue.
```
diff --git a/Esm/ib/src/linux/startup/Makefile b/Esm/ib/src/linux/startup/Makefile
index 86bf29b..4001ac8 100644
--- a/Esm/ib/src/linux/startup/Makefile
+++ b/Esm/ib/src/linux/startup/Makefile
@@ -143,7 +143,7 @@ include $(TL_DIR)/Makerules/Maketargets.build
 include $(TL_DIR)/Makerules/Maketargets.install
 
 $(BUILDDIR)/opafmd: $(BUILDDIR)/opafmd.o
-	$(VS)$(CC) -pthread $(LDFLAGS)$@ $(BUILDDIR)/opafmd.o $(LDLIBS)
+	$(VS)$(CC) -pie -pthread $(LDFLAGS)$@ $(BUILDDIR)/opafmd.o $(LDLIBS)
 
 $(BUILDDIR)/opaxmlextract: $(PROJ_BIN_DIR)/opaxmlextract
 	cp $(PROJ_BIN_DIR)/opaxmlextract $(BUILDDIR)
diff --git a/Esm/ib/src/linux/startup/fm_config/Makefile b/Esm/ib/src/linux/startup/fm_config/Makefile
index cbf277f..99b3635 100644
--- a/Esm/ib/src/linux/startup/fm_config/Makefile
+++ b/Esm/ib/src/linux/startup/fm_config/Makefile
@@ -150,10 +150,10 @@ include $(TL_DIR)/Makerules/Maketargets.build
 include $(TL_DIR)/Makerules/Maketargets.install
 
 $(BUILDDIR)/config_check: $(BUILDDIR)/config_check.o
-	$(VS)$(CC) $(LDFLAGS)$@ $(BUILDDIR)/config_check.o $(LDLIBS)
+	$(VS)$(CC) -pie $(LDFLAGS)$@ $(BUILDDIR)/config_check.o $(LDLIBS)
 
 $(BUILDDIR)/config_convert: $(BUILDDIR)/config_convert.o
-	$(VS)$(CC) $(LDFLAGS)$@ $(BUILDDIR)/config_convert.o $(LDLIBS)
+	$(VS)$(CC) -pie $(LDFLAGS)$@ $(BUILDDIR)/config_convert.o $(LDLIBS)
 
 # install for stage phase
 #include $(TL_DIR)/Makerules/Maketargets.stage
diff --git a/Makerules/Maketargets.build b/Makerules/Maketargets.build
index ebf9fdf..9cd7844 100755
--- a/Makerules/Maketargets.build
+++ b/Makerules/Maketargets.build
@@ -127,7 +127,7 @@ else # BUILD_TARGET_OS is not VXWORKS
 $(CMD_TARGETS_EXE) :	$(CT_OBJECTS) $(RSCOBJECTS) $(DEPLIBS_TARGETS)
 	@echo Linking executable...
 	@mkdir -p $(dir $@)
-	$(VS)$(CC) $(LDFLAGS)$@ $(CT_OBJECTS) $(LDLIBS)
+	$(VS)$(CC) -pie $(LDFLAGS)$@ $(CT_OBJECTS) $(LDLIBS)
 endif # BUILD_TARGET_OS is VXWORKS
 endif # CMD_TARGETS_EXE non-null
```

**Comment by sjb017** [comment]
I applied this patch to the tip of master, which has been updated with patches from our latest release, and I get build errors.

**Comment by HonggangLI** [comment]
Could you please share the details of the build errors?

**Comment by sjb017** [comment]
It wants the objects to have been compiled with -fPIC:

Linking executable...
gcc -pie  -Wl,-z,noexecstack -z relro -z now     -L/nfs/site/home/sbreyer/rpmbuild/BUILD/opa-fm-10.7.0.0/builtlibs.VIEO_HOST.debug -Xlinker -rpath-link -Xlinker /nfs/site/home/sbreyer/rpmbuild/BUILD/opa-fm-10.7.0.0/builtlibs.VIEO_HOST.debug  -o rm_version build.VIEO_HOST.debug/rm_version.o -Xlinker --start-group   -Xlinker --end-group -lpthread
/usr/bin/ld: build.VIEO_HOST.debug/rm_version.o: relocation R_X86_64_32 against `.rodata.str1.1' can not be used when making a shared object; recompile with -fPIC
build.VIEO_HOST.debug/rm_version.o: could not read symbols: Bad value
collect2: error: ld returned 1 exit status
make[4]: *** [rm_version] Error 1
make[4]: Leaving directory `/nfs/site/home/sbreyer/rpmbuild/BUILD/opa-fm-10.7.0.0/MakeTools/rm_version'
make[3]: *** [CMDS] Error 2
make[3]: Leaving directory `/nfs/site/home/sbreyer/rpmbuild/BUILD/opa-fm-10.7.0.0/MakeTools'
make[2]: *** [buildall] Error 2
make[2]: Leaving directory `/nfs/site/home/sbreyer/rpmbuild/BUILD/opa-fm-10.7.0.0/MakeTools'
make[1]: *** [tools] Error 2
make[1]: Leaving directory `/nfs/site/home/sbreyer/rpmbuild/BUILD/opa-fm-10.7.0.0/Esm'
make: *** [buildall] Error 2

**Comment by sjb017** [comment]
I have, however, been able to get a build working in our native repository using -fPIC and -pie, so I can run tests for performance. We would like to analyze to make sure there is not a significant degradation in the performance of our executables when built this way.

**Comment by sjb017** [comment]
In our test build using -fPIC and -pie, our tests indicate a performance reduction ranging from 2% in one test to 40% in the worst case. We suspect that we would continue to see the reduction in the range of 2-4% and so are reluctant to include this build option in our product.

**Comment by ddalessa** [comment]
RHEL RPM package has been using -fPIC for quite some time without issues. We will roll that in to our 10.14.5 release in the future. For now it is available for 10.12.1.0.6 for Fedora.
